### PR TITLE
Rework URL path manipulation APIs to better match results of Foundation

### PR DESCRIPTION
**Pull request by ky-is**
The current Skip URL path manipulation APIs are pretty basic and sometimes produce significantly different results from Foundation. Some of these differences are incorrect (i.e. not detecting the `pathExtension` of a URL with a query string), and some are implementation choices (Foundation's `pathComponents` include `"/"` prefix/suffix components under certain conditions). The goal of all of these changes is simply to better match the output of the equivalent Foundation APIs.

Affected methods:
- `String.deletingLastPathComponent` (this is on NSString in Foundation, should it actually be here?)
- ~`URL.path(percentEncoded:) -> String?`~ `URL.path(percentEncoded:) -> String`
- `URL.pathComponents`
- `URL.appendingPathComponent`
- `URL.appendingPathExtension`
- `URL.deletingLastPathComponent`
- `URL.deletingPathExtension`
- `URLComponents.percentEncodedPath` as the internal canonical path reference

I referenced implementations from Swift Foundation so I added the license reference preamble to a couple files as well. Not sure how correct/necessary that is here. I generally took a middle ground of adapting the existing vs replacing with the Foundation implementation.

I created some bespoke tests for these functions as `test_URLStrings` cannot transpile currently and I wasn't sure how to go about that. They should illustrate ways that the current Skip implementation differs from Foundation. Note that I did not attempt to handle all the edge cases that Foundation does (i.e. relative path components, see commented test cases for `deletingLastPathComponent`). It would take more time to understand these cases and I didn't have personal need for them (and they weren't previously handled regardless).

Presumably this PR will need some discussion/changes before it is ready to merge, or a different approach like starting from scratch with the Foundation implementations may be desirable.

- [x] REQUIRED: I have signed the [Contributor Agreement](https://github.com/skiptools/clabot-config)
- [x] REQUIRED: I have tested my change locally with `swift test`
- [x] OPTIONAL: I have tested my change on an iOS simulator or device
- [x] OPTIONAL: I have tested my change on an Android emulator or device

### CODE REVIEW

**Comment by ky-is** [comment]
I could use help interpreting this CI error, forgive me I am still a Skip noob. `swift test` passes for me on my machine so I'm having trouble tracking down the cause. This test seems to fail in the Mac environment where it should be using the builtin Foundation version and not the Skip implementation correct?

File: Tests/SkipFoundationTests/Network/TestURL.swift [hunk 1]
```diff
@@ -781,6 +781,87 @@ class TestURL : XCTestCase {
...
+    }
+
+    func test_deletingPathExtension() {
+        let urlsExpected = [
+            ("https://www.swift.org/", "https://www.swift.org/"),
+            ("https://www.swift.org/.", "https://www.swift.org/."),
+            ("https://www.swift.org/a.", "https://www.swift.org/a."),
+            ("https://www.swift.org/a..b", "https://www.swift.org/a."),
+            ("https://www.swift.org/.a.b", "https://www.swift.org/.a"),
+            ("https://www.swift.org/..b", "https://www.swift.org/..b"),
```
Comment (aabewhite) [inline Tests/SkipFoundationTests/Network/TestURL.swift:835]: This fails on macOS 14.5 for me. I believe our CI machines are also an even older version. It also fails on iOS 17. But it passes on iOS 18, and are you on macOS 15? If it's passing for you on macOS 15, then it's clearly a versioning issue. Up to you whether you want to deal with trying to update the test for #if SKIP and if #available or just omit this case

**Review by aabewhite — CHANGES_REQUESTED** [review]
A couple of minor comments, but otherwise looks great - thanks!

File: Sources/SkipFoundation/URL.swift [hunk 2]
```diff
@@ -293,15 +307,42 @@ public struct URL : Hashable, CustomStringConvertible, Codable, KotlinConverting
...
+            end = path[end...].firstIndex(of: "/") ?? path.endIndex
+            if start != end {
+                let subpath = String(path[start..<end])
+                result.append(subpath)
+            }
+            start = path[end...].firstIndex { $0 != "/" } ?? path.endIndex
+            end = start
+        }
+
+        if path.count > 2 && path.hasSuffix("//") { //TODO
```
[the remark this answers is not in the document]
Comment (ky-is) [reply Sources/SkipFoundation/URL.swift]: I intended this to be a placeholder but I didn't find a better implementation today. Added a comment!

File: Tests/SkipFoundationTests/Network/TestURL.swift (hunk 1 above)
Comment (ky-is) [reply Tests/SkipFoundationTests/Network/TestURL.swift]: Ahh thank you Abe. I did not even consider that these APIs would still be changing.

I'm commenting out the test case since there are doubtless other weird edge cases like this.

Comment (marcprux) [reply Tests/SkipFoundationTests/Network/TestURL.swift]: Looks like the Robolectric CI is passing now, but there is a [new failure](https://github.com/skiptools/skip-foundation/actions/runs/12361630109/job/34499204528#step:16:881) on the Android emulator testing step:

```
12-16 22:04:54.177  2175  2193 I TestRunner: started: testLocalizableStrings$SkipFoundation_debugAndroidTest(skip.foundation.LocaleTests)
12-16 22:04:54.266  2175  2193 E TestRunner: failed: testLocalizableStrings$SkipFoundation_debugAndroidTest(skip.foundation.LocaleTests)
12-16 22:04:54.267  2175  2193 E TestRunner: ----- begin exception -----
12-16 22:04:54.267  2175  2193 E TestRunner: java.lang.NullPointerException
12-16 22:04:54.267  2175  2193 E TestRunner: 	at skip.foundation.URL.deletingLastPathComponent(URL.kt:428)
12-16 22:04:54.267  2175  2193 E TestRunner: 	at skip.foundation.LocaleTests.testLocalizableStrings$SkipFoundation_debugAndroidTest(LocaleTests.kt:305)
12-16 22:04:54.267  2175  2193 E TestRunner: 	at java.lang.reflect.Method.invoke(Native Method)
```

I haven't dug into this, but I suspect the problem may be related to a localization strings URL: the `Localizable.strings` file it is a stored in a traditional Java URL resource. In Robolectric, this will be a plain `file:///` URL (since Robolectric doesn't package things up into an apk), but when deployed in the Android environment, it will be something like `jar:file:/data/app/~~GrNJyKuGMG-gs4i97rlqHg==/skip.ui.test-5w0MhfIK6rNxUpG8yMuXgg==/base.apk!/skip/ui/Resources/Localizable.strings`. I suspect that this might be tripping up some of the new URL logic in this PR.

Ideally you could replicate this by testing against an Android emulator (by setting the `ANDROID_SERIAL=emulator-5554` environment variable, either in Xcode's Scheme run arguments or from the command-line). You could also just paste the URL above into your test case to see if `deletingLastPathComponent()` fails on it.

Comment (ky-is) [reply Tests/SkipFoundationTests/Network/TestURL.swift]: Seems java.net.URI doesn't like the `jar:file:` scheme. It returns null for `rawPath` and `path` (but `toString` still works which is why the old implementation worked (Swift.URL's methods also don't work with these URLs).

I'm not sure what approach to take, I see a couple options:
- Fall back to the old implementation when `rawPath` returns null (should be the least disruptive if this behavior is relied upon elsewhere in the Skip ecosystem, but it's kind of ugly)
- Special case the `jar:file:` URLs in SkipFoundation.URL
- Use the old implementation as custom helpers for processing these URLs in local tests (not sure if other Skip libraries rely on the old behavior of the URL methods?)

Comment (aabewhite) [reply Tests/SkipFoundationTests/Network/TestURL.swift]: I don't think jar:file: URLs are entirely uncommon on Android, so that would rule out the last option. 

Do you have any particular strategy in mind for special-casing jar:file: URLs? E.g. trim any "jar:" prefix when performing operations and then add it back?

Comment (ky-is) [reply Tests/SkipFoundationTests/Network/TestURL.swift]: I tried `android.net.Uri` and it returns null paths as well. `java.net.URL` works because it returns `file:...` as part of the path (technically incorrect but enough for these test cases). So potentially swapping to URL is an option? But I don't know what additional side effects that would cause.

Are we sure this is useful as part of the public Skip API? I'm not enough of an Android developer to know how common interacting with raw jar paths is.

Removing the `jar:` prefix is enough for everything to work normally so that's definitely an option. It could be done individually in each method, or the constructor setting a `isJAR` flag and removing the prefix. It doesn't seem ideal to run this logic on every URL to support one special case, but of course I'm deferring to you.

Comment (marcprux) [reply Tests/SkipFoundationTests/Network/TestURL.swift]: > Are we sure this is useful as part of the public Skip API? I'm not enough of an Android developer to know how common interacting with raw jar paths is.

All resources are stored in the jars, so any calls like `Bundle.module.url(forResource: "image", withExtension: "png")` are going to return a `jar:file:` URL.

Comment (marcprux) [reply Tests/SkipFoundationTests/Network/TestURL.swift]: The issue is coming is from the use of `URLComponents`, right? I wonder if there could be some preprocessing and postprocessing of the URL before being passed through that that would remove all but the final `scheme` from the URL before creating the `URLComponents`, and then adding them back when extracting the URL back out from it.

Some thing like this pseudo-code:

```swift
func withURLComponents(f: (URLComponents) -> URI) -> URI {
    var schemes: [String] = []
    var withSingleSchemeURI = self.uri
    if let scheme = withSingleSchemeURI.scheme {
        schemes.append(scheme)
        withSingleSchemeURI = URI(withSingleSchemeURI.substring(scheme.count + 1))
    }
    // add back the final scheme so we have at least one
    if let scheme = schemes.dropLast() {
        withSingleSchemeURI = URI(scheme + ":" + withSingleSchemeURI.absoluteString())
    }

    let components = URLComponents(withSingleSchemeURI)
    var transformed = f(components) // perform the operation in the single-scheme URI URLComponents

    // now add back the deleted schemes
    for scheme in schemes.reversed() {
        transformed = URI(scheme + ":" + transformed.absoluteString())        
    }

    return transformed
}
```

Comment (ky-is) [reply Tests/SkipFoundationTests/Network/TestURL.swift]: > All resources are stored in the jars, so any calls like `Bundle.module.url(forResource: "image", withExtension: "png")` are going to return a `jar:file:` URL.

Ah right.. I'm getting confused by the platform implementations not supporting their own urls but I guess they expect you to be using abstractions.

> The issue is coming is from the use of `URLComponents`, right?

The core of the issue is that `path`/`rawPath` return null for non-standard uris. Which then prevents URLComponents from working.

Considering none of the platform versions support schemes with multiple colons, imho we should special case these `jar:` urls instead of supporting any generic multi-colon scheme. I took a stab at this in the latest commit, though I modified URL rather than URLComponents so that all URL methods will work. But it certainly could be done within URLComponents as well if that's more desirable.

Comment (marcprux) [reply Tests/SkipFoundationTests/Network/TestURL.swift]: Unfortunately, I don't think we can do [this](https://github.com/skiptools/skip-foundation/pull/35/files#diff-70e4edf2e14052846e6fee0c53f315422bb0d1ed3cb0e63c915130a5b8e6c995R132-R135):

```
    public init?(string: String, encodingInvalidCharacters: Bool) {
        let string = string.hasPrefix("jar:file:") ? "jarfile" + string.dropFirst(8) : string
        do {
            self.platformValue = java.net.URI(string) // throws on malformed
```

The problem is that we need the underlying `platformValue` to always be a valid (i.e., loadable) value. For example, Kotlin bridging will be using that underlying value directly when transforming between a `skip.foundation.URL` and a `java.net.URI`, and the scheme `jarfile:` will not be valid. So if Kotlin does `url.platformValue.toURL().openConnection()`, it will fail.

It occurs to me that the easiest solution might be to have [URLComponents](https://github.com/skiptools/skip-foundation/blob/main/Sources/SkipFoundation/URLComponents.swift) tolerate double schemes (like "jar:file:"). I'm not sure where/why it is currently failing, but a test case with a "jar:file:" scheme would probably identify it.

Comment (ky-is) [reply Tests/SkipFoundationTests/Network/TestURL.swift]: Gotcha. I switched to handling this in URLComponents, which means URL params like `path` and `pathComponents` will return empty for these jar urls (exactly the old behavior before this branch). I changed the helper methods to remove their dependency on `platformValue.rawPath` as a result, and added explicit test cases for manipulating these jar urls.

Comment (marcprux) [reply Tests/SkipFoundationTests/Network/TestURL.swift]: Looks good, thanks!

Comment (ky-is) [reply Tests/SkipFoundationTests/Network/TestURL.swift]: Thanks both for all your assistance! Took me a while to get my head around this one.
